### PR TITLE
Esp8266 Part 2

### PR DESCRIPTION
**Pull request by fabaff**
This part is about MQTT and MicroPython. Work in progress...

Part 1 #532

### CODE REVIEW

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+>>> resp1 = urequests.get(url, headers=headers)
+>>> print(resp1.json())
+{'name': 'Sensor2', 'value': 23}
+```
+
+I think that with this piece of information you should be able to use the [Home Assistant RESTful API](/developers/rest_api/).
+
+Now, MQTT. You can find simple examples for publishing and subscribing for MQTT in the [MicroPython](https://github.com/micropython/micropython-lib) library overview in the section for [umqtt](https://github.com/micropython/micropython-lib/tree/master/umqtt.simple). 
+
+[@davea](https://github.com/davea) created [sonoff-mqtt](https://github.com/davea/sonoff-mqtt). This code will work on ESP8622 based devices too. The example below is adopted from the work of [@davea](https://github.com/davea) as we don't want to re-invent the wheel.
```
Comment (balloob) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: Is this the one that is distributed with MicroPython (the stretch goal) ?

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+```bash
+$ curl -X POST -d "value=23" http://192.168.1.18:5000/sensor2
+```
+
+The conversion of the `curl` command to use with `urequests` is very simple. Basically it's a one-liner but splitted to be easier to read.
+ 
+```python
+>>> import urequests
+>>> url = 'http://192.168.1.18:5000/sensor2'
```
Comment (balloob) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: Why not have the example write a state to Home Assistant ?

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+```
+
+```yaml
+sensor:
+  - platform: mqtt
+    state_topic: "home/esp8266_[last part of the MAC address]"
+    name: "MicroPython"
+```
+
+One disadvantage of the ESP8266 boards is that there is only one analog input. If you are using DHT11, Dallas 18S20, or other sensors you can work out this limitation.
```
Comment (balloob) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: I think that this line feels a bit out of place.

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+>>> print(resp.json())
+{'name': 'Sensor2', 'value': 23}
+>>> resp1 = urequests.get(url, headers=headers)
+>>> print(resp1.json())
+{'name': 'Sensor2', 'value': 23}
+```
+
+I think that with this piece of information you should be able to use the [Home Assistant RESTful API](/developers/rest_api/).
+
+Now, MQTT. You can find simple examples for publishing and subscribing for MQTT in the [MicroPython](https://github.com/micropython/micropython-lib) library overview in the section for [umqtt](https://github.com/micropython/micropython-lib/tree/master/umqtt.simple).
```
Comment (balloob) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: Let's add a linkable_title above this?

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown (hunk 1 above)
Comment (fabaff) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: I don't think that sonoff is/was a stretch goal of the campaign. @davea created it around umqtt which is/was a stretch goal.

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown (hunk 2 above)
Comment (fabaff) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: Don't wanted to spoil the fun...makes more sense to communicate with HA.

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown (hunk 3 above)
Comment (fabaff) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: It is. Thanks

File: source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown (hunk 4 above)
Comment (fabaff) [inline source/_posts/2016-07-25-esp8266-and-micropython-part2.markdown]: Sure

**Comment by fabaff** [comment]
Need to figure out the issue with `urequests.post`.

**Comment by sfam** [comment]
yes, if header line starts with "Location":
https://github.com/micropython/micropython-lib/blob/master/urllib.urequest/urllib/urequest.py#L57-L58

From [API docs](https://home-assistant.io/developers/rest_api/#post-apistatesltentity_id):

> 
Return code is 200 if the entity existed, 201 if the state of a new entity was set. **A location header will be returned** with the url of the new resource. The response body will contain a JSON encoded State object.

which is found here: 
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/api.py#L205

not sure how to workaround this...

**Pull request closed without merging**